File: packages/contracts/src/mcp.ts
import type { PlanMode, ProviderKind } from "./domain";

export type RecommendedSequenceItemType = "skill" | "tool";
export type SkillArtifactType = "text" | "json" | "file" | "command" | "url";

export interface SessionAttachInput {
  provider: ProviderKind;
  providerSessionId?: string;
  cwd: string;
}

export interface SessionAttachOutput {
  sessionId: string;
  taskId: string;
  mode: PlanMode;
  activeStepId?: string;
  activeSkillProfileId?: string;
  recommendedNextCalls: string[];
}

export interface ContextGetCurrentInput {
  sessionId: string;
}

export interface ContextTaskView {
  id: string;
  title: string;
  prompt: string;
  goal?: string;
  constraints?: string[];
}

export interface ContextWorkspaceView {
  id: string;
  rootPath: string;
}

export interface ContextWorktreeView {
  id: string;
  path: string;
  branchName?: string;
}

export interface ContextSessionView {
  id: string;
  provider: ProviderKind;
  state: string;
}

export interface ContextGetCurrentOutput {
  task: ContextTaskView;
  workspace: ContextWorkspaceView;
  worktree?: ContextWorktreeView;
  session: ContextSessionView;
}

export interface RecommendedSequenceItem {
  type: RecommendedSequenceItemType;
  name: string;
}

export interface TaskGetNextActionInput {
  sessionId: string;
}

export interface TaskGetNextActionStep {
  id: string;
  title: string;
  description?: string;
  successCriteria?: string[];
  leaseToken?: string;
}

export interface TaskGetNextActionOutput {
  mode: PlanMode;
  step?: TaskGetNextActionStep;
  activeSkills: string[];
  recommendedSequence?: RecommendedSequenceItem[];
}

export interface TaskUpdateProgressInput {
  sessionId: string;
  stepId?: string;
  leaseToken?: string;
  summary: string;
  details?: Record<string, unknown>;
}

export interface McpAcceptedOutput {
  accepted: true;
}

export interface TaskCompleteStepInput {
  sessionId: string;
  stepId: string;
  leaseToken?: string;
  summary?: string;
  outputs?: Record<string, unknown>;
}

export interface TaskCompleteStepOutput {
  accepted: true;
  nextStepId?: string;
}

export interface TaskBlockStepInput {
  sessionId: string;
  stepId: string;
  reason: string;
}

export interface SkillListActiveInput {
  sessionId: string;
}

export interface SkillListActiveOutput {
  activeSkills: string[];
  preferredSkills?: string[];
  forbiddenSkills?: string[];
}

export interface SkillInvokeInput {
  sessionId: string;
  stepId?: string;
  skill: string;
  input: Record<string, unknown>;
}

export interface SkillArtifact {
  type: SkillArtifactType;
  uri?: string;
  text?: string;
}

export interface SkillInvokeOutput {
  summary: string;
  result?: Record<string, unknown>;
  artifacts?: SkillArtifact[];
}

File: crates/codebar-contracts/src/mcp.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::domain::{PlanMode, ProviderKind};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RecommendedSequenceItemType {
    Skill,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SessionAttachInput {
    pub provider: ProviderKind,
    pub provider_session_id: Option<String>,
    pub cwd: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SessionAttachOutput {
    pub session_id: String,
    pub task_id: String,
    pub mode: PlanMode,
    pub active_step_id: Option<String>,
    pub active_skill_profile_id: Option<String>,
    pub recommended_next_calls: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContextGetCurrentInput {
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContextTaskView {
    pub id: String,
    pub title: String,
    pub prompt: String,
    pub goal: Option<String>,
    pub constraints: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContextWorkspaceView {
    pub id: String,
    pub root_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContextWorktreeView {
    pub id: String,
    pub path: String,
    pub branch_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContextSessionView {
    pub id: String,
    pub provider: ProviderKind,
    pub state: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ContextGetCurrentOutput {
    pub task: ContextTaskView,
    pub workspace: ContextWorkspaceView,
    pub worktree: Option<ContextWorktreeView>,
    pub session: ContextSessionView,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RecommendedSequenceItem {
    pub r#type: RecommendedSequenceItemType,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskGetNextActionInput {
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskGetNextActionStep {
    pub id: String,
    pub title: String,
    pub description: Option<String>,
    pub success_criteria: Option<Vec<String>>,
    pub lease_token: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskGetNextActionOutput {
    pub mode: PlanMode,
    pub step: Option<TaskGetNextActionStep>,
    pub active_skills: Vec<String>,
    pub recommended_sequence: Option<Vec<RecommendedSequenceItem>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TaskUpdateProgressInput {
    pub session_id: String,
    pub step_id: Option<String>,
    pub lease_token: Option<String>,
    pub summary: String,
    pub details: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AcceptedOutput {
    pub accepted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TaskCompleteStepInput {
    pub session_id: String,
    pub step_id: String,
    pub lease_token: Option<String>,
    pub summary: Option<String>,
    pub outputs: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskCompleteStepOutput {
    pub accepted: bool,
    pub next_step_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskBlockStepInput {
    pub session_id: String,
    pub step_id: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SkillListActiveInput {
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SkillListActiveOutput {
    pub active_skills: Vec<String>,
    pub preferred_skills: Option<Vec<String>>,
    pub forbidden_skills: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillInvokeInput {
    pub session_id: String,
    pub step_id: Option<String>,
    pub skill: String,
    pub input: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SkillArtifactType {
    Text,
    Json,
    File,
    Command,
    Url,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillArtifact {
    pub r#type: SkillArtifactType,
    pub uri: Option<String>,
    pub text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillInvokeOutput {
    pub summary: String,
    pub result: Option<Value>,
    pub artifacts: Option<Vec<SkillArtifact>>,
}
